client/__tests__/e2e/data.persistence.e2e.test.ts

```diff
@@ -0,0 +1,543 @@
+/* eslint-disable no-await-in-loop -- await in loop is needed to emulate sequential user actions  */
+import { test } from "@chromatic-com/playwright";
+import crypto from "crypto";
+import fs from "fs/promises";
+import path from "path";
+import { existsSync } from "fs";
+
+import mockSetup from "./playwright.global.setup";
+
+import {
+  calcDragCoordinates,
+  drag,
+  snapshotTestGraph,
+  goToPage,
+  createCategory,
+  addLabelToCategoryWithSelection,
+  assertCategoryExists,
+  assertCategoryDoesNotExist,
+  assertLabelExists,
+  assertLabelDoesNotExist,
+  deleteCategory,
+  renameCategory,
+  deleteLabel,
+  renameLabel,
+} from "./cellxgeneActions";
+import { APP_URL_BASE } from "../common/constants";
+
+const { describe } = test;
+
+// Test constants
+const TEST_CATEGORY_NAME = "persistence_test_category";
+const TEST_LABEL_NAME = "persistence_test_label";
+const SECOND_LABEL_NAME = "persistence_second_label";
+const RENAMED_CATEGORY_NAME = "renamed_persistence_category";
+const RENAMED_LABEL_NAME = "renamed_persistence_label";
+
+// Use persistence test URL with /w/ dataroot
+const DATASET = "pbmc3k.cxg";
+const persistenceTestURL = [APP_URL_BASE, "w", DATASET, ""].join("/");
+
+const PROJECT_ROOT =
+  process.env.PROJECT_ROOT ??
+  (existsSync(path.join(process.cwd(), "example-dataset"))
+    ? process.cwd()
+    : path.resolve(process.cwd(), ".."));
+
+// TODO #754
+test.beforeEach(mockSetup);
+
+const SIDE_PANEL = "layout-graph-side";
+
+const graphInstanceTestIds = ["layout-graph", "layout-graph-side"];
+
+// Helper function to generate unique user ID for each test
+function generateUniqueUserId(): string {
+  return `test-user-${crypto.randomBytes(8).toString("hex")}`;
+}
+
+// Helper function to set unique user ID header for authentication
+function setUserIdHeader(page: any, userId: string) {
+  return page.setExtraHTTPHeaders({
+    "X-Auth-Request-User": userId,
+  });
+}
+
+// Helper function to wait for autosave to complete
+async function waitForAutosaveComplete(page: any) {
+  // Wait for autosave element to exist
+  await page.waitForSelector("#autosave", { timeout: 10000 });
+
+  // Wait for at least 2.5s to ensure the autosave timer has had a chance to run
+  // The timer runs every 2.5s, so we need to wait at least that long
+  await page.waitForTimeout(2500);
+
+  // Now wait for autosave to complete - it should show "autosave-complete"
+  await page.waitForSelector('#autosave[data-testclass="autosave-complete"]', {
+    timeout: 8000, // Allow time for network request + processing after timer fires
+  });
+
+  // Give a small buffer to ensure state is stable
+  await page.waitForTimeout(500);
+}
+
+// Helper function to refresh page and wait for load
+async function refreshPageAndWait(page: any) {
+  await page.reload({ waitUntil: "networkidle" });
+  await page.waitForTimeout(2000); // Wait for app initialization
+}
+
+// Helper function to cleanup user data after test
```

Comment:
The cleanup hook down here should keep example-dataset tidy between runs; just flagging that rdev paths will need the same layout.